### PR TITLE
Auto Provisioning Kubernetes LoadBalancer Services with F5

### PR DESCRIPTION
**Pull request by arbhoj**

### CODE REVIEW

File: content/posts/f5-k8s-lb-svc/index.md [hunk 1]
```diff
@@ -0,0 +1,262 @@
+---
+authors: ["ADD AUTHOR NAME"]
+title: "Auto Provisioning Kubernetes LoadBalancer Services with F5"
+date: 2022-11-17T00:35:56-05:00
+featured: false
+tags: [loadbalancer]
+excerpt: Learn how to configure a kubernetes cluster to provision a service of type LoadBalancer with F5 BIG-IP
```
Comment (jimmidyson) [inline content/posts/f5-k8s-lb-svc/index.md:7]:
```suggestion
excerpt: Learn how to configure a Kubernetes cluster to provision a service of type LoadBalancer with F5 BIG-IP 
```

File: content/posts/f5-k8s-lb-svc/index.md [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+authors: ["ADD AUTHOR NAME"]
+title: "Auto Provisioning Kubernetes LoadBalancer Services with F5"
+date: 2022-11-17T00:35:56-05:00
+featured: false
+tags: [loadbalancer]
+excerpt: Learn how to configure a kubernetes cluster to provision a service of type LoadBalancer with F5 BIG-IP 
+feature_image: feature.png
+---
+
+# Options to provision LoadBalancer Services in an on-premise Kubernetes Cluster
```
Comment (jimmidyson) [inline content/posts/f5-k8s-lb-svc/index.md:11]:
```suggestion
# Options to provision LoadBalancer Services in an on-premise Kubernetes cluster
```

File: content/posts/f5-k8s-lb-svc/index.md [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+
+When deploying Kubernetes clusters in the cloud (AWS, Azure, GCP etc.) a Kubernetes service of type [LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) is by default provisioned automatically by the native load-balancer service (e.g. [ELB](https://aws.amazon.com/elasticloadbalancing/) on AWS). It's not the same for on-premise clusters, which require additional components to be deployed/configured to get this functionality. There are few Kubernetes native options like [metallb](https://metallb.universe.tf/) and [kube-vip](https://kube-vip.io/). However, for production clusters and clusters that process heavy traffic, an external load-balancer might be a better choice. [F5 BIG-IP LTM](https://www.f5.com/products/big-ip-services/local-traffic-manager) is one of the most widely used load-balancers in the industry today and the best thing is that it integrates really nicely with Kubernetes.
```
Comment (jimmidyson) [inline content/posts/f5-k8s-lb-svc/index.md:13]:
```suggestion
When deploying Kubernetes clusters in the cloud (AWS, Azure, GCP etc.) a Kubernetes service of type [LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) is by default provisioned automatically by the appropriate [cloud controller manager](https://kubernetes.io/docs/concepts/architecture/cloud-controller/) using the native load-balancer service (e.g. [ELB](https://aws.amazon.com/elasticloadbalancing/) on AWS). It's not the same for on-premise clusters, which require additional components to be deployed/configured to get this functionality. There are a few Kubernetes native options like [metallb](https://metallb.universe.tf/) and [kube-vip](https://kube-vip.io/). However, for production clusters and clusters that process heavy traffic, an external load-balancer might be a better choice. [F5 BIG-IP LTM](https://www.f5.com/products/big-ip-services/local-traffic-manager) is one of the most widely used load-balancers in the industry today and the best thing is that it integrates really nicely with Kubernetes. 
```

File: content/posts/f5-k8s-lb-svc/index.md [hunk 4]
```diff
@@ -0,0 +1,262 @@
...
+
+When deploying Kubernetes clusters in the cloud (AWS, Azure, GCP etc.) a Kubernetes service of type [LoadBalancer](https://kubernetes.io/docs/concepts/services-networking/service/#loadbalancer) is by default provisioned automatically by the native load-balancer service (e.g. [ELB](https://aws.amazon.com/elasticloadbalancing/) on AWS). It's not the same for on-premise clusters, which require additional components to be deployed/configured to get this functionality. There are few Kubernetes native options like [metallb](https://metallb.universe.tf/) and [kube-vip](https://kube-vip.io/). However, for production clusters and clusters that process heavy traffic, an external load-balancer might be a better choice. [F5 BIG-IP LTM](https://www.f5.com/products/big-ip-services/local-traffic-manager) is one of the most widely used load-balancers in the industry today and the best thing is that it integrates really nicely with Kubernetes. 
+
+In this blog we will see how to configure an on-premise Kubernetes cluster to integrate with [F5 BIG-IP LTM](https://www.f5.com/products/big-ip-services/local-traffic-manager) to provision a service of type LoadBalancer.
```
Comment (jimmidyson) [inline content/posts/f5-k8s-lb-svc/index.md:15]:
```suggestion
In this blog we will see how to configure an on-premise Kubernetes cluster to integrate with [F5 BIG-IP LTM](https://www.f5.com/products/big-ip-services/local-traffic-manager) to provision a service of type `LoadBalancer`.
```

File: content/posts/f5-k8s-lb-svc/index.md [hunk 5]
```diff
@@ -0,0 +1,262 @@
...
+# Integrating a Kubernetes cluster with F5
+
+## Requirements
+1. Pre-configured F5 BIG cluster 
+2. F5 Partition that will be managed by this automation along with credentials for a service account that has admin permissions for the given partition
+3. AS3 3.39 or newer installed on the F5 cluster
+4. IP's availble to be used as VIPs
+5. Working Kubernetes cluster with PV Storage configured
+6. If using the CAPI steps then a CAPI bootstrap/management cluster and configurations to deploy a cluster. Refer [DKP](https://docs.d2iq.com/dkp/latest/infrastructure-quick-start-guides) documentation for more details.
+7. This uses `docker.io/f5networks/f5-ipam-controller:0.1.5` & `docker.io/f5networks/k8s-bigip-ctlr:2.9.1` images. Download, retag and push the images to a local registry and change the deployment spec to point to a local image registry for airgapped environments.
```
Comment (jimmidyson) [inline content/posts/f5-k8s-lb-svc/index.md:26]: Can you please add links to the different components here? I don't know what they are and not every reader will either so linking would help the reader.

File: content/posts/f5-k8s-lb-svc/index.md [hunk 6]
```diff
@@ -0,0 +1,262 @@
...
+
+>Versions used to test:
+>- BIG-IP: 16.1.3.1 Build 0.0.11 Point Release 1
+>- AS3: v3.39.0
+>- BIG-IP-CTLR: 2.9.1
```
Comment (jimmidyson) [inline content/posts/f5-k8s-lb-svc/index.md:31]:
```suggestion
This blog used the following versions to test:

* BIG-IP: 16.1.3.1 Build 0.0.11 Point Release 1
* AS3: v3.39.0
* BIG-IP-CTLR: 2.9.1
```

File: content/posts/f5-k8s-lb-svc/index.md [hunk 7]
```diff
@@ -0,0 +1,262 @@
...
+
+## Steps to Integrate a Kuberntes Cluster with F5 BIG-IP
```
Comment (jimmidyson) [inline content/posts/f5-k8s-lb-svc/index.md:33]:
```suggestion
## Steps to Integrate a Kuberntes cluster with F5 BIG-IP
```

File: content/posts/f5-k8s-lb-svc/index.md [hunk 8]
```diff
@@ -0,0 +1,262 @@
...
+Here are the steps to be performed for a kubernetes cluster that is to be integrated with F5 Big IP to provision services of type LoadBalancer. 
+
+There are two options based on whether F5 controllers are to be directly deployed to the target cluster or deployed via ClusterResourceSets for a CAPI provisioned cluster either at cluter creation time or after the cluster has been deployed.
```
Comment (jimmidyson) [inline content/posts/f5-k8s-lb-svc/index.md:36]:
```suggestion
There are two options based on whether F5 controllers are to be directly deployed to the target cluster, or deployed via [`ClusterResourceSets`](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-resource-set.html) for a [CAPI](https://cluster-api.sigs.k8s.io/) provisioned cluster either at cluster creation time or after the cluster has been deployed.
```

**Review by jimmidyson — CHANGES_REQUESTED** [review]
Sorry I didn't realise this post wasn't finished when I started reviewing. I'm going to mark it as a draft PR to indicate that now. Please set it back to ready for review when it's ready.

**Comment by arbhoj** [comment]
Accepted review comments and added the required links in the requirements section

**Comment by arbhoj** [comment]
Merging

File: content/posts/f5-k8s-lb-svc/index.md [hunk 9]
```diff
@@ -0,0 +1,261 @@
...
+This blog used the following versions to test:
+
+* BIG-IP: 16.1.3.1 Build 0.0.11 Point Release 1
+* AS3: v3.39.0
+* BIG-IP-CTLR: 2.9.1
+
+## Steps to Integrate a Kuberntes cluster with F5 BIG-IP
+There are two options based on whether F5 controllers are to be directly deployed to the target cluster, or deployed via [`ClusterResourceSets`](https://cluster-api.sigs.k8s.io/tasks/experimental-features/cluster-resource-set.html) for a [CAPI](https://cluster-api.sigs.k8s.io/) provisioned cluster either at cluster creation time or after the cluster has been deployed.
+
+<br/>
```
Comment (tillt) [inline content/posts/f5-k8s-lb-svc/index.md:37]: Do we really need those? Lemme assume we don't and suggest changes...

Comment (tillt) [inline content/posts/f5-k8s-lb-svc/index.md:37]:
```suggestion

```

File: content/posts/f5-k8s-lb-svc/index.md [hunk 10]
```diff
@@ -0,0 +1,261 @@
...
+
+### Option 1: Directly deploy F5 Controllers to a Kubernetes Cluster
+<br/>
```
Comment (tillt) [inline content/posts/f5-k8s-lb-svc/index.md:40]:
```suggestion

```

File: content/posts/f5-k8s-lb-svc/index.md [hunk 11]
```diff
@@ -0,0 +1,261 @@
...
+```
+export F5_USER=f5-user
+export F5_PASSWD=f5-password
+export KUBECONFIG=kubeconfig-file-path
+
+kubectl create secret generic f5-bigip-ctlr-login -n kube-system --from-literal=username=${F5_USER} --from-literal=password=${F5_PASSWD} 
+
+helm install -f f5-${CLUSTER_NAME}-values.yaml f5ctlr f5-stable/f5-bigip-ctlr --version 0.0.21
+```
+<br/>
```
Comment (tillt) [inline content/posts/f5-k8s-lb-svc/index.md:95]:
```suggestion

```

File: content/posts/f5-k8s-lb-svc/index.md [hunk 12]
```diff
@@ -0,0 +1,261 @@
...
+  - ReadWriteOnce
+  resources:
+    requests:
+      storage: 0.1Gi
+EOF
+
+helm install -f f5-ipam-${CLUSTER_NAME}-values.yaml f5-ipam  f5-ipam-stable/f5-ipam-controller --version 0.0.1
+```
+<br/>
+<br/>
```
Comment (tillt) [inline content/posts/f5-k8s-lb-svc/index.md:156]:
```suggestion

```

File: content/posts/f5-k8s-lb-svc/index.md [hunk 13]
```diff
@@ -0,0 +1,261 @@
...
+
+### Option 2: Deploy Automatically via CAPI 
+>Note: If deploying to a CAPI provisioned Kubernetes Cluster like [DKP](https://docs.d2iq.com/dkp/latest/infrastructure-quick-start-guides) instead of running the install command manually, the above can be packaged into a CAPI ClusterResourceSet by doing the following and incorporated into the cluster deployment process.
+
+<br/>
```
Comment (tillt) [inline content/posts/f5-k8s-lb-svc/index.md:161]:
```suggestion

```

**Review by tillt — COMMENTED** [review]
Looks great to me.

Just learned that we will need to find a nice picture representing this post graphically. Checking the offerings of https://unsplash.com/ following @jimmidyson nice suggestion.

**Comment by jimmidyson** [comment]
Rebased...

**Review by jimmidyson — APPROVED** [review]

**Review by tillt — APPROVED** [review]

**Review by tillt — DISMISSED** [review]

**Review by tillt — DISMISSED** [review]

**Review by vacas5 — DISMISSED** [review]

**Review by tillt — APPROVED** [review]

**Comment by tillt** [comment]
The preview keeps showing the last article from Kim-Norman - that is confusing. @jimmidyson any idea?

**Pull request merged**